### ISSUE TITLE
[BUG/ISSUE] Parallelization or precision issue in full-chemistry

### ISSUE THREAD

**Issue by yantosca** [geoschem/geos-chem#1157]
### What institution are you from?
GCST

### Description of the problem
While looking into issues #1103 and https://github.com/geoschem/HEMCO/issues/132, I discovered what appears to be a further parallelization issue affecting full-chemistry simulations.  Specifically, we seem to get different values of O3 (and potentially other species) at a particular grid box when running on a different number of cores.

I had noticed that the Inorganic Iodine HEMCO extension was returning very small differences in a couple of grid boxes (~ 1e-22) when using the same gcclassic executable on 5 cores and on 8 cores.  Because this emission uses the O3 concentration at the surface, I found that we got identical results if we shut chemistry off.  Thus this points to either a parallelization or precision issue in chemistry.

I put some debug output to print the value of O3 at 4 x 5 box (25,9,1) after the first integration step.

On 1 core:
```console
 >>> C(ind_O3) @ 25,9,1 before integrate:    670937484717.53247     
 >>> C(ind_O3) @ 25,9,1 after integrate :    670932428474.80713     
```

On 5 cores:
```console
 >>> C(ind_O3) @ 25,9,1 before integrate:    670937484717.53247     
 >>> C(ind_O3) @ 25,9,1 after integrate :    670932428476.99219  
```

On 8 cores:
```console
 >>> C(ind_O3) @ 25,9,1 before integrate:    670937484717.53247     
 >>> C(ind_O3) @ 25,9,1 after integrate :    670932428472.16040    
```
As you can see, the O3 concentration is identical before the call to KPP on all core counts but the O3 concentration after the first integration step all are different.  We are continuing to investigate this.

**Comment by yantosca** [comment]
The fix has now been pushed to commit 72c8866ee68a60639be743255bfad2454357af7e.  However, we are still investigating what appears to be very small differences in the ship emissions:

```console
Using configuration file hco58.yml
... Printing totals and differences
... Only showing variables with |absolute difference| > 0.0
Variable               Ref=5 cores              Dev=8 cores              Dev - Ref
EmisNO_Ship          : 1.5393503e-10          | 1.5393503e-10          | -2.144715586069626e-30 
InvPARANOX_NO        : 1.5393503e-10          | 1.5393503e-10          | -2.144715586069626e-30 
EmisNO_Total         : 1.9958073e-08          | 1.9958073e-08          | -2.144715586069626e-30 
```
will investigate further.

**Comment by yantosca** [comment]
See issue https://github.com/geoschem/HEMCO/issues/133 for more about the parallelization issue in PARANOX.

**Comment by yantosca** [comment]
As noted in https://github.com/geoschem/HEMCO/issues/133

> I can confirm that this issue is not being caused by PARANOX per se, but it is an artifact of small differences that have been creeping in from GEOS-Chem operations.
> 
> In commit [geoschem/geos-chem@5bd68df](https://github.com/geoschem/geos-chem/commit/5bd68df14cf5e78717abbd687f6821bc6561a766), I added several fixes for numerical stability to `GeosCore/tpcore_fvdas_mod.F90`. Basically making sure that arrays are zeroed before they are used, as well as any arguments declared with `INTENT(OUT)`.
> 
> I am looking into a similar issue in the VDIFF PBL mixing module. More to come.

**Comment by yantosca** [comment]
I have isolated the source of the differences to this DO loop in routine `WET_SETTLING` in `GeosCore/seasalt_mod.F90`.  It might be that we need to add an `!$OMP REDUCTION` for the summation. clause.  Will investigate further.

Offending DO loop:

https://github.com/geoschem/geos-chem/blob/04024de551f76bc2c7edda43e88f2b3eeaa62471/GeosCore/seasalt_mod.F90#L660-L693

**Comment by yantosca** [comment]
Remaining paralllelization issues have been solved with PR #1190.  We now obtain identical results regardless of the number of cores used.  We can close this issue now.
